File: ReSharper.FSharp/src/FSharp.ExternalFormatter/src/CodeFormatterProvider.cs
using Fantomas;
using FSharp.Compiler;
using FSharp.Compiler.CodeAnalysis;
using FSharp.Compiler.Diagnostics;
using FSharp.Compiler.SourceCodeServices;
using FSharp.Compiler.Text;
using JetBrains.Rider.FSharp.ExternalFormatter.Client;
using Microsoft.FSharp.Collections;
using Microsoft.FSharp.Control;

namespace FSharp.ExternalFormatter
{
  internal interface ICodeFormatterProvider
  {
    string FormatSelection(RdFormatSelectionArgs args);
    string FormatDocument(RdFormatDocumentArgs args);
  }

  internal class BundledCodeFormatter : ICodeFormatterProvider
  {
    private readonly FSharpChecker myChecker =
      FSharpChecker.Create(null, null, null, null, null, null, null, null, null, null);

    private readonly FormatConfig.FormatConfig myDefaultFormatConfig = FormatConfig.FormatConfig.Default;

    public string FormatSelection(RdFormatSelectionArgs args) =>
      FSharpAsync.StartAsTask(
          CodeFormatter.FormatSelectionAsync(
            args.FileName,
            Convert(args.Range),
            SourceOrigin.SourceOrigin.NewSourceString(args.Source),
            Convert(args.FormatConfig),
            Convert(args.ParsingOptions),
            myChecker), null, null)
        .Result;

    public string FormatDocument(RdFormatDocumentArgs args) =>
      FSharpAsync.StartAsTask(
          CodeFormatter.FormatDocumentAsync(
            args.FileName,
            SourceOrigin.SourceOrigin.NewSourceString(args.Source),
            Convert(args.FormatConfig),
            Convert(args.ParsingOptions),
            myChecker), null, null)
        .Result;

    private static Range Convert(RdRange range) =>
      CodeFormatter.MakeRange(range.FileName, range.StartLine, range.StartCol, range.EndLine, range.EndCol);

    private static FSharpParsingOptions Convert(RdParsingOptions options) =>
      new FSharpParsingOptions(
        options.SourceFiles,
        FSharpList<string>.Empty,
        FSharpDiagnosticOptions.Default,
        false,
        options.LightSyntax,
        false,
        false);

    private FormatConfig.FormatConfig Convert(RdFormatConfig config) =>
      new FormatConfig.FormatConfig(
        config.IndentSize,
        config.MaxLineLength,
        myDefaultFormatConfig.SemicolonAtEndOfLine,
        config.SpaceBeforeParameter,
        config.SpaceBeforeLowercaseInvocation,
        config.SpaceBeforeUppercaseInvocation,
        config.SpaceBeforeClassConstructor,
        config.SpaceBeforeMember,
        config.SpaceBeforeColon,
        config.SpaceAfterComma,
        config.SpaceBeforeSemicolon,
        config.SpaceAfterSemicolon,
        config.IndentOnTryWith,
        config.SpaceAroundDelimiter,
        config.MaxIfThenElseShortWidth,
        config.MaxInfixOperatorExpression,
        config.MaxRecordWidth,
        myDefaultFormatConfig.MaxRecordNumberOfItems,
        myDefaultFormatConfig.RecordMultilineFormatter,
        config.MaxArrayOrListWidth,
        myDefaultFormatConfig.MaxArrayOrListNumberOfItems,
        myDefaultFormatConfig.ArrayOrListMultilineFormatter,
        config.MaxValueBindingWidth,
        config.MaxFunctionBindingWidth,
        myDefaultFormatConfig.MaxDotGetExpressionWidth,
        config.MultilineBlockBracketsOnSameColumn,
        config.NewlineBetweenTypeDefinitionAndMembers,
        config.KeepIfThenInSameLine,
        config.MaxElmishWidth,
        config.SingleArgumentWebMode,
        config.AlignFunctionSignatureToIndentation,
        config.AlternativeLongMemberDefinitions,
        myDefaultFormatConfig.MultiLineLambdaClosingNewline,
        myDefaultFormatConfig.DisableElmishSyntax, myDefaultFormatConfig.EndOfLine,
        myDefaultFormatConfig.StrictMode);
  }
}

File: ReSharper.FSharp/src/FSharp.Psi.Features/src/Formatter/CodeFormatterProvider.fs
namespace JetBrains.ReSharper.Plugins.FSharp.Services.Formatter

open System
open FSharp.Compiler.CodeAnalysis
open FSharp.Compiler.Text
open FSharp.ExternalFormatter.Protocol
open JetBrains.Lifetimes
open JetBrains.ProjectModel
open JetBrains.Rd.Tasks
open JetBrains.Rider.FSharp.ExternalFormatter.Server

[<SolutionComponent>]
type CodeFormatterProvider(solution: ISolution, externalFormatterFactory: ExternalFormatterProcessFactory) =
    let mutable connection: ExternalFormatterConnection = null

    let isConnectionAlive () =
        isNotNull connection && connection.IsActive

    let connect () =
        if isConnectionAlive () then () else
        let formatterHostLifetime = Lifetime.Define(solution.GetLifetime())
        connection <- externalFormatterFactory.Create(formatterHostLifetime.Lifetime).Run()

    let execute (action: unit -> string) =
        connect ()
        connection.Execute(action)

    let convertRange (range: range) =
        RdRange(range.FileName, range.StartLine, range.StartColumn, range.EndLine, range.EndColumn)

    let convertFormatSettings (settings: FSharpFormatSettingsKey) =
        RdFormatConfig
            (settings.INDENT_SIZE, settings.WRAP_LIMIT, settings.SpaceBeforeParameter,
             settings.SpaceBeforeLowercaseInvocation, settings.SpaceBeforeUppercaseInvocation,
             settings.SpaceBeforeClassConstructor, settings.SpaceBeforeMember, settings.SpaceBeforeColon,
             settings.SpaceAfterComma, settings.SpaceBeforeSemicolon, settings.SpaceAfterSemicolon,
             settings.IndentOnTryWith, settings.SpaceAroundDelimiter, settings.MaxIfThenElseShortWidth,
             settings.MaxInfixOperatorExpression, settings.MaxRecordWidth, settings.MaxArrayOrListWidth,
             settings.MaxValueBindingWidth, settings.MaxFunctionBindingWidth,
             settings.MultilineBlockBracketsOnSameColumn, settings.NewlineBetweenTypeDefinitionAndMembers,
             settings.KeepIfThenInSameLine, settings.MaxElmishWidth, settings.SingleArgumentWebMode,
             settings.AlignFunctionSignatureToIndentation, settings.AlternativeLongMemberDefinitions)

    let convertParsingOptions (options: FSharpParsingOptions) =
        let lightSyntax =
            match options.LightSyntax with
            | Some x -> Nullable<bool> x
            | None -> Nullable<bool>()

        RdParsingOptions(options.SourceFiles, lightSyntax)

    member x.FormatSelection(filePath: string, range: range, source: string, settings: FSharpFormatSettingsKey,
                             options: FSharpParsingOptions) =
        let args = RdFormatSelectionArgs(filePath, convertRange range, source, convertFormatSettings settings,
                                         convertParsingOptions options)

        execute (fun () -> connection.ProtocolModel.FormatSelection.Sync(args, RpcTimeouts.Maximal))

    member x.FormatDocument(filePath: string, source: string, settings: FSharpFormatSettingsKey,
                            options: FSharpParsingOptions) =
        let args = RdFormatDocumentArgs(filePath, source, convertFormatSettings settings, convertParsingOptions options)

        execute (fun () -> connection.ProtocolModel.FormatDocument.Sync(args, RpcTimeouts.Maximal))
